flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java

```diff
@@ -71,12 +74,26 @@
 
 	protected abstract B getStateBackend() throws Exception;
 
+	protected abstract long getManagedMemorySize();
```

Comment:
I don't really like that we introduce this method for all sub classes of this class. Instead, it should only be relevant for RocksDB related tests. One alternative would be to introcude

```
protected void modifyMockEnvironmentBuilder(MockEnvironmentBuilder builder) {}
```

which we call in `before()` and which can be overriden by the RocksDB tests to set the managed memory size. That way, all non RocksDB related test classes would not need to implement this method.

Reply:
The original `MockEnvironmentBuilder` would already have a default `memorySize` as `1024 * 32 * 1024`. The reason I introduced a `getManagedMemorySize` method here is to let non-RocksDB test could not reserve any off-heap memory.
If you agree, I think we could use original `MockEnvironmentBuilder` directly but also reserve off-heap memory for non-RocksDB tests.

Reply:
Ah I understand now. You actually want to not give managed memory to all state backend tests which don't need it. If we want to make sure that these tests actually work without `MemorySegments`, I can see the benefit of disabling managed memory for them. Maybe we could then do it the other way around meaning that we only explicitly deactivate/set managed memory to `0` for those tests but by default the `MockEnvironmentBuilder` gives you 32 MB of managed memory.

Reply:
Previously, memory manager for test would reserve on-heap memory by default. However, after [FLINK-15023](https://issues.apache.org/jira/browse/FLINK-15023), on-heap memory is not managed anymore. 
As I talked with @xintongsong offline, the reason did not remove on-heap memory from memory manager for testing is mainly due to the consideration for tests could be finished more quickly if we only allocate on-heap memory for testing.
In this PR, we enable rocksDB to use managed memory by default. And I'll correct those places not setting off-heap managed memory. This might lead to some places which not need off-heap also reserved off-heap memory. 
However, I found too many places to cut that redundant off-heap memory to polish the tests to the best situation and a bit out of scope of this PR targets. After I think about it for a while, I prefer to let the part to remove unnecessary off-heap memory for testing as another story or issue but not included in this PR. If so, I'll not call to set managed memory as `0` in these mem state backends. 
What do you think of this?

Reply:
Not sure I fully understand.

All managed memory (on heap and off heap) is not eagerly allocated or reserved. Having some managed memory by default would not hurt, because it is not materialized as memory segments. All tests that would not use it would not behave any different if the managed memory was there or not.

Reply:
Sounds good to me! +1 for this approach.

Reply:
@StephanEwen I think @xintongsong might give some explanations here.
If the configured off-heap memory would not hurt, my approach works.
If the configured off-heap has a bit impact, my approach should still be picked to let flowing-up commits to fix.
In a nut shell, I'll update this with my approach before I go to sleep tonight.

Reply:
@StephanEwen
I find it difficult to follow the entire conversation, but I can try to explain what's the problem of having on-heap managed memory for RocksDB state backend test cases: 
- RocksDB state backend (or to be specific `RocksDBOperationUtils#allocateSharedCachesIfConfigured`) calls `MemoryManager#getSharedMemoryResourceForManagedMemory` for reserving and allocating managed memory, with 1.0 as the `fractionToInitializeWith`.
- In `getSharedMemoryResourceForManagedMemory`, we first compute the memory size with the fraction 1.0, then reserve that amount of memory.
- While `computeMemorySize` returns memory size including both on-heap and off-heap, we reserve only off-heap memory.
- If the memory manager contains both on-heap and off-heap memory, no matter eagerly allocated/reserved or not, we will try to reserve more off-heap memory than it has.
- E.g., memory manager contains 50mb on-heap budget and 100mb off-heap budget, in `getSharedMemoryResourceForManagedMemory` we will first compute the memory size as 150mb, then try to reserve 150mb off-heap memory which is larger than the off-heap budget.

Currently memory manager created from the testing util class `MemoryManagerBuilder` will always contain some on-heap memory, which is added in the constructor of `MemoryManagerBuilder`. When @Myasuka asked yesterday offline, I suggested to check whether any budget is set in `MemoryManagerBuilder#build`, and only add the default on-heap budget if no other budget is set. In this way, we can keep the batch test cases to use on-heap managed memory for less testing overhead, while have RocksDB state backend test cases to explicitly use off-heap managed memory only.

@tillrohrmann @Myasuka 
I'm not sure whether its necessary to set managed memory to 0 for non-RocksDB cases. Since managed memory is always lazily allocated/reserved, I think streaming test cases with non-RocksDB state backend should not reserve managed memory at all, and batch test cases should only allocate on-heap managed memory if needed (which is the same behavior as before).

Reply:
Thanks for the clarification. So to sum it up, it should be ok to configure off-heap managed memory for streaming test cases. Moreover, there is no harm in giving non-RocksDB test cases a bit of managed memory. If, then this should be changed follow-ups. Last but not least, batch test cases will still use on-heap managed memory.